cps/kni.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <rte_ethdev.h>
+#include <rte_kni.h>
+#include <rte_malloc.h>
+
+#include "gatekeeper_cps.h"
+#include "gatekeeper_main.h"
+#include "kni.h"
+
+#define INSMOD_CMD "insmod "
+#define KNI_KMOD "rte_kni"
+#define RMMOD_CMD ("rmmod " KNI_KMOD)
+
+extern long init_module(void *, unsigned long, const char *);
+extern long delete_module(const char *, unsigned int);
+
+int
+kni_change_if(uint8_t port_id, uint8_t if_up)
+{
+	return (if_up)
+		? rte_eth_dev_set_link_up(port_id)
+		: rte_eth_dev_set_link_down(port_id);
+}
+
+int
+kni_change_mtu(uint8_t port_id, unsigned new_mtu)
+{
+	if (unlikely(new_mtu > ETHER_MAX_LEN))
+		return -EINVAL;
+
+	return rte_eth_dev_set_mtu(port_id, new_mtu);
+}
+
+/*
+ * Inserting and removing modules.
+ *
+ * This code is adapted for use in DPDK from the source code of
+ * insmod and rmmod in the module-init-tools package.
+ */
+
+static void *
+grab_file(const char *filename, unsigned long *size)
+{
+	unsigned int max = 16384;
```

Comment:
Please add the following comment at the end of the line to make this number less magic: `/* 16KB */`